rpc/client.go

```diff
@@ -26,212 +26,606 @@ type Transport interface {
 	Close() error
 }
 
-type Client interface {
-	Logger() log.Logger
-	ServerConfig() ServerConfig
-	User() discord.User
-	V() int
-	Transport() Transport
-
-	Subscribe(event Event, args CmdArgs, handler Handler) error
-	Unsubscribe(event Event, args CmdArgs) error
-	Send(message Message, handler Handler) error
-	Close()
+type responseMessage struct {
+	data MessageData
+	err  error
 }
 
 func NewClient(clientID snowflake.ID, opts ...ConfigOpt) (Client, error) {
 	config := DefaultConfig()
 	config.Apply(opts)
 
-	client := &clientImpl{
-		logger:          config.Logger,
+	client := Client{
+		Logger:          config.Logger,
 		eventHandlers:   map[Event]Handler{},
-		commandHandlers: map[string]internalHandler{},
+		commandChannels: map[string]chan responseMessage{},
 		readyChan:       make(chan struct{}, 1),
+		clientID:        clientID,
 	}
 
 	if config.Transport == nil {
 		var err error
 		config.Transport, err = config.TransportCreate(clientID, config.Origin)
 		if err != nil {
-			return nil, err
+			return Client{}, err
 		}
 	}
-	client.transport = config.Transport
+	client.Transport = config.Transport
+
+	return client, nil
+}
 
-	go client.listen(client.transport)
+type Client struct {
+	Logger    log.Logger
+	Transport Transport
+
+	eventHandlers   map[Event]Handler
+	commandChannels map[string]chan responseMessage
+	clientID        snowflake.ID
+
+	readyChan    chan struct{}
+	User         discord.User
+	ServerConfig ServerConfig
+	V            int
+}
+
+func (c *Client) Open() error {
+	go c.listen(c.Transport)
 
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
 	select {
 	case <-ctx.Done():
-		return nil, ctx.Err()
-	case <-client.readyChan:
+		return ctx.Err()
+	case <-c.readyChan:
 	}
+	return nil
+}
 
-	return client, nil
+func (c *Client) sendBytes(r io.Reader) error {
+	writer, err := c.Transport.NextWriter()
+	if err != nil {
+		return err
+	}
+	defer writer.Close()
+
+	buff := new(bytes.Buffer)
+	newWriter := io.MultiWriter(writer, buff)
+
+	_, err = io.Copy(newWriter, r)
+	if err != nil {
+		return err
+	}
+
+	data, _ := io.ReadAll(buff)
+	c.Logger.Tracef("Sending message: data: %s", string(data))
+
+	return err
 }
 
-type clientImpl struct {
-	logger    log.Logger
-	transport Transport
+func (c *Client) Authorize(scopes []discord.OAuth2Scope, rpcToken string, username string) (string, error) {
+	args := CmdArgsAuthorize{
+		ClientID: c.clientID,
+		Scopes:   scopes,
+		RPCToken: rpcToken,
+		Username: username,
+	}
 
-	eventHandlers   map[Event]Handler
-	commandHandlers map[string]internalHandler
+	if res, err := c.send(Message{
+		Cmd:  CmdAuthorize,
+		Args: args,
+	}); err != nil {
+		return "", err
+	} else {
```

Comment:
`else` = smelly

Reply:
Resolved in [4ef8ffc](https://github.com/dennisrijsdijk/disgo/commit/4ef8ffc03ba58f70cb22951e155ddf66e18ef4e6)